File: sdk-android/src/com/mobileapptracker/MATUrlBuilder.java
package com.mobileapptracker;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Date;
import java.util.UUID;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.util.Log;

class MATUrlBuilder {
    private static MATParameters params;
    /**
     * Builds a new link string based on parameter values.
     * @return encrypted URL string based on class settings.
     */
    public static String buildLink(MATEvent eventData, MATPreloadData preloaded, boolean debugMode) {
        params = MATParameters.getInstance();
        
        StringBuilder link = new StringBuilder("https://").append(params.getAdvertiserId()).append(".");
        if (debugMode) {
            link.append(MATConstants.MAT_DOMAIN_DEBUG);
        } else {
            link.append(MATConstants.MAT_DOMAIN);
        }
        
        link.append("/serve?ver=").append(params.getSdkVersion());
        link.append("&transaction_id=").append(UUID.randomUUID().toString());

        safeAppend(link, "sdk", "android");
        safeAppend(link, "action", params.getAction());
        safeAppend(link, "advertiser_id", params.getAdvertiserId());
        safeAppend(link, "package_name", params.getPackageName());
        safeAppend(link, "referral_source", params.getReferralSource());
        safeAppend(link, "referral_url", params.getReferralUrl());
        safeAppend(link, "site_id", params.getSiteId());
        safeAppend(link, "tracking_id", params.getTrackingId());
        
        if (eventData.getEventId() != 0) {
            safeAppend(link, "site_event_id", Integer.toString(eventData.getEventId()));
        }
        if (!params.getAction().equals("session")) {
            safeAppend(link, "site_event_name", eventData.getEventName());
        }

        // Append preloaded params, must have attr_set=1 in order to attribute
        if (preloaded != null) {
            link.append("&attr_set=1");
            safeAppend(link, "publisher_id", preloaded.publisherId);
            safeAppend(link, "offer_id", preloaded.offerId);
            safeAppend(link, "publisher_ref_id", preloaded.publisherReferenceId);
            safeAppend(link, "publisher_sub_publisher", preloaded.publisherSubPublisher);
            safeAppend(link, "publisher_sub_site", preloaded.publisherSubSite);
            safeAppend(link, "publisher_sub_campaign", preloaded.publisherSubCampaign);
            safeAppend(link, "publisher_sub_adgroup", preloaded.publisherSubAdgroup);
            safeAppend(link, "publisher_sub_ad", preloaded.publisherSubAd);
            safeAppend(link, "publisher_sub_keyword", preloaded.publisherSubKeyword);
            safeAppend(link, "advertiser_sub_publisher", preloaded.advertiserSubPublisher);
            safeAppend(link, "advertiser_sub_site", preloaded.advertiserSubSite);
            safeAppend(link, "advertiser_sub_campaign", preloaded.advertiserSubCampaign);
            safeAppend(link, "advertiser_sub_adgroup", preloaded.advertiserSubAdgroup);
            safeAppend(link, "advertiser_sub_ad", preloaded.advertiserSubAd);
            safeAppend(link, "advertiser_sub_keyword", preloaded.advertiserSubKeyword);
            safeAppend(link, "publisher_sub1", preloaded.publisherSub1);
            safeAppend(link, "publisher_sub2", preloaded.publisherSub2);
            safeAppend(link, "publisher_sub3", preloaded.publisherSub3);
            safeAppend(link, "publisher_sub4", preloaded.publisherSub4);
            safeAppend(link, "publisher_sub5", preloaded.publisherSub5);
        }

        // If allow duplicates on, skip duplicate check logic
        String allowDups = params.getAllowDuplicates();
        if (allowDups != null) {
            int intAllowDups = Integer.parseInt(allowDups);
            if (intAllowDups == 1) {
                link.append("&skip_dup=1");
            }
        }

        // If logging on, use debug mode
        if (debugMode) {
            link.append("&debug=1");
        }
        
        return link.toString();
    }
    

    /**
     * Builds data in conversion link based on class member values, to be encrypted.
     * @return URL-encoded string based on class settings.
     */
    public static synchronized String buildDataUnencrypted(MATEvent eventData) {
        params = MATParameters.getInstance();

        StringBuilder link = new StringBuilder();

        link.append("connection_type=" + params.getConnectionType());
        safeAppend(link, "age", params.getAge());
        safeAppend(link, "altitude", params.getAltitude());
        safeAppend(link, "android_id", params.getAndroidId());
        safeAppend(link, "android_id_md5", params.getAndroidIdMd5());
        safeAppend(link, "android_id_sha1", params.getAndroidIdSha1());
        safeAppend(link, "android_id_sha256", params.getAndroidIdSha256());
        safeAppend(link, "app_ad_tracking", params.getAppAdTrackingEnabled());
        safeAppend(link, "app_name", params.getAppName());
        safeAppend(link, "app_version", params.getAppVersion());
        safeAppend(link, "app_version_name", params.getAppVersionName());
        safeAppend(link, "country_code", params.getCountryCode());
        safeAppend(link, "device_brand", params.getDeviceBrand());
        safeAppend(link, "device_carrier", params.getDeviceCarrier());
        safeAppend(link, "device_cpu_type", params.getDeviceCpuType());
        safeAppend(link, "device_cpu_subtype", params.getDeviceCpuSubtype());
        safeAppend(link, "device_model", params.getDeviceModel());
        safeAppend(link, "device_id", params.getDeviceId());
        safeAppend(link, "existing_user", params.getExistingUser());
        safeAppend(link, "facebook_user_id", params.getFacebookUserId());
        safeAppend(link, "gender", params.getGender());
        safeAppend(link, "google_aid", params.getGoogleAdvertisingId());
        safeAppend(link, "google_ad_tracking_disabled", params.getGoogleAdTrackingLimited());
        safeAppend(link, "google_user_id", params.getGoogleUserId());
        safeAppend(link, "insdate", params.getInstallDate());
        safeAppend(link, "installer", params.getInstaller());
        safeAppend(link, "install_referrer", params.getInstallReferrer());
        safeAppend(link, "is_paying_user", params.getIsPayingUser());
        safeAppend(link, "language", params.getLanguage());
        safeAppend(link, "last_open_log_id", params.getLastOpenLogId());
        safeAppend(link, "latitude", params.getLatitude());
        safeAppend(link, "longitude", params.getLongitude());
        safeAppend(link, "mac_address", params.getMacAddress());
        safeAppend(link, "mat_id", params.getMatId());
        safeAppend(link, "mobile_country_code", params.getMCC());
        safeAppend(link, "mobile_network_code", params.getMNC());
        safeAppend(link, "open_log_id", params.getOpenLogId());
        safeAppend(link, "os_version", params.getOsVersion());
        safeAppend(link, "sdk_plugin", params.getPluginName());
        safeAppend(link, "android_purchase_status", params.getPurchaseStatus());
        safeAppend(link, "referrer_delay", params.getReferrerDelay());
        safeAppend(link, "screen_density", params.getScreenDensity());
        safeAppend(link, "screen_layout_size", params.getScreenWidth() + "x" + params.getScreenHeight());
        safeAppend(link, "sdk_version", params.getSdkVersion());
        safeAppend(link, "truste_tpid", params.getTRUSTeId());
        safeAppend(link, "twitter_user_id", params.getTwitterUserId());
        safeAppend(link, "conversion_user_agent", params.getUserAgent());
        safeAppend(link, "user_email_md5", params.getUserEmailMd5());
        safeAppend(link, "user_email_sha1", params.getUserEmailSha1());
        safeAppend(link, "user_email_sha256", params.getUserEmailSha256());
        safeAppend(link, "user_id", params.getUserId());
        safeAppend(link, "user_name_md5", params.getUserNameMd5());
        safeAppend(link, "user_name_sha1", params.getUserNameSha1());
        safeAppend(link, "user_name_sha256", params.getUserNameSha256());
        safeAppend(link, "user_phone_md5", params.getPhoneNumberMd5());
        safeAppend(link, "user_phone_sha1", params.getPhoneNumberSha1());
        safeAppend(link, "user_phone_sha256", params.getPhoneNumberSha256());
        
        // Append event-level params
        safeAppend(link, "attribute_sub1", eventData.getAttribute1());
        safeAppend(link, "attribute_sub2", eventData.getAttribute2());
        safeAppend(link, "attribute_sub3", eventData.getAttribute3());
        safeAppend(link, "attribute_sub4", eventData.getAttribute4());
        safeAppend(link, "attribute_sub5", eventData.getAttribute5());
        safeAppend(link, "content_id", eventData.getContentId());
        safeAppend(link, "content_type", eventData.getContentType());
        // Event-level currency overrides MAT class-level
        if (eventData.getCurrencyCode() != null) {
            safeAppend(link, "currency_code", eventData.getCurrencyCode());
        } else {
            safeAppend(link, "currency_code", params.getCurrencyCode());
        }
        if (eventData.getDate1() != null) {
            safeAppend(link, "date1", Long.toString(eventData.getDate1().getTime()/1000));
        }
        if (eventData.getDate2() != null) {
            safeAppend(link, "date2", Long.toString(eventData.getDate2().getTime()/1000));
        }
        if (eventData.getLevel() != 0) {
            safeAppend(link, "level", Integer.toString(eventData.getLevel()));
        }
        if (eventData.getQuantity() != 0) {
            safeAppend(link, "quantity", Integer.toString(eventData.getQuantity()));
        }
        if (eventData.getRating() != 0) {
            safeAppend(link, "rating", Double.toString(eventData.getRating()));
        }
        safeAppend(link, "search_string", eventData.getSearchString());
        safeAppend(link, "advertiser_ref_id", eventData.getRefId());
        safeAppend(link, "revenue", Double.toString(eventData.getRevenue()));
        
        return link.toString();
    }

    
    /**
     * Update the Google Ad ID and install referrer, if present, and encrypts the data string.
     * @return encrypted string
     */
    public static synchronized String updateAndEncryptData(String data, MATEncryption encryption) {
        StringBuilder updatedData = new StringBuilder(data);
        
        params = MATParameters.getInstance();
        if (params != null) {
            String gaid = params.getGoogleAdvertisingId();
            if (gaid != null && !data.contains("&google_aid=")) {
                safeAppend(updatedData, "google_aid", gaid);
                safeAppend(updatedData, "google_ad_tracking_disabled", params.getGoogleAdTrackingLimited());
            }
            
            String androidId = params.getAndroidId();
            if (androidId != null && !data.contains("&android_id=")) {
                safeAppend(updatedData, "android_id", androidId);
            }
            
            String referrer = params.getInstallReferrer();
            if (referrer != null && !data.contains("&install_referrer=")) {
                safeAppend(updatedData, "install_referrer", referrer);
            }
            String userAgent = params.getUserAgent();
            if (userAgent != null && !data.contains("&conversion_user_agent=")) {
                safeAppend(updatedData, "conversion_user_agent", userAgent);
            }
        }
        // Add system date of original request
        if (!data.contains("&system_date=")) {
            long now = new Date().getTime()/1000;
            safeAppend(updatedData, "system_date", Long.toString(now));
        }
        
        String updatedDataStr = updatedData.toString();
        try {
            updatedDataStr = MATEncryption.bytesToHex(encryption.encrypt(updatedDataStr));
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        return updatedDataStr;
}
    
    /**
     * Builds JSONObject for body of POST request
     * @return appropriately parameterized object
     */
    public static synchronized JSONObject buildBody(JSONArray eventItems, String iapData, String iapSignature, JSONArray emails) {
        JSONObject postData = new JSONObject();

        try {
            if (eventItems != null) {
                postData.put("data", eventItems);
            }
            if (iapData != null) {
                postData.put("store_iap_data", iapData);
            }
            if (iapSignature != null) {
                postData.put("store_iap_signature", iapSignature);
            }
            if (emails != null) {
                postData.put("user_emails", emails);
            }
        } catch (JSONException e) {
            Log.d(MATConstants.TAG, "Could not build JSON body of request");
            e.printStackTrace();
        }
        
        return postData;
    }
    
    /*
     * URL builders
     */
    private static synchronized void safeAppend(StringBuilder link, String key, String value) {
        if (value != null && !value.equals("")) {
            try {
                link.append("&" + key + "=" + URLEncoder.encode(value, "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                Log.w(MATConstants.TAG, "failed encoding value " + value + " for key " + key); 
                e.printStackTrace();
            }
        }
    }
}

File: sdk-windows/MATWindows8.1-SDK/MATWindows8.1-SDK/MATUrlBuilder.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Text;


namespace MobileAppTracking
{
    public class MATUrlBuilder
    {
        private const string DOMAIN = "engine.mobileapptracking.com";
        private const string IV = "heF9BATUfWuISyO8";
        private const string SDK_TYPE = "windows";
        private const string SDK_VERSION = "3.3";
        
        //TODO: Look into using JSON rather than strings
        public static string BuildUrl(string action, string eventName, double revenue, string currency, string refId, List<MATEventItem> eventItems, Parameters parameters)
        {
            StringBuilder url = new StringBuilder("https://");
            url.Append(Uri.EscapeUriString(parameters.advertiserId)).Append(".");

            if (parameters.DebugMode)
                url.Append("debug.");
            url.Append(DOMAIN).Append("/serve?sdk=").Append(SDK_TYPE).Append("&ver=").Append(SDK_VERSION);
            url.Append("&advertiser_id=").Append(Uri.EscapeUriString(parameters.advertiserId));
            url.Append("&mat_id=").Append(Uri.EscapeUriString(parameters.MatId));
            url.Append("&action=").Append(Uri.EscapeUriString(action));
            url.Append("&package_name=").Append(Uri.EscapeUriString(parameters.PackageName));
            url.Append("&transaction_id=").Append(Guid.NewGuid().ToString().ToUpper());
            // Append event name/ID for events
            if (action.Equals("conversion"))
            {
                long value;
                if (long.TryParse(eventName, out value))
                    url.Append("&site_event_id=").Append(eventName);
                else
                    url.Append("&site_event_name=").Append(Uri.EscapeUriString(eventName));
            }

            // Append open log id
            if (parameters.OpenLogId != null)
                url.Append("&open_log_id=").Append(Uri.EscapeUriString(parameters.OpenLogId));
            if (parameters.LastOpenLogId != null)
                url.Append("&last_open_log_id=").Append(Uri.EscapeUriString(parameters.LastOpenLogId));

            if (parameters.AllowDuplicates)
                url.Append("&skip_dup=1");
            if (parameters.DebugMode)
                url.Append("&debug=1");
            if (parameters.ExistingUser)
                url.Append("&existing_user=1");

            // Construct encrypted data params and append to url
            StringBuilder data = new StringBuilder();
            // Add UNIX timestamp as system date
            long timestamp = UnixTimestamp();
            data.Append("&system_date=").Append(timestamp.ToString());

            data.Append("&windows_aid=").Append(Uri.EscapeUriString(parameters.WindowsAid));
            data.Append("&app_name=").Append(Uri.EscapeUriString(parameters.AppName));
            data.Append("&app_version=").Append(Uri.EscapeUriString(parameters.AppVersion));
            data.Append("&os_id=").Append(Uri.EscapeUriString(parameters.ASHWID));
            if (parameters.UserAgent != null)
                data.Append("&user_agent=").Append(Uri.EscapeUriString(parameters.UserAgent));
            if (parameters.AppAdTracking)
                data.Append("&app_ad_tracking=1");
            else
                data.Append("&app_ad_tracking=0");

            if (revenue > 0)
                data.Append("&revenue=").Append(Uri.EscapeUriString(revenue.ToString(parameters.culture)));
            if (currency != null)
                data.Append("&currency_code=").Append(Uri.EscapeUriString(currency));
            if (refId != null)
                data.Append("&advertiser_ref_id=").Append(Uri.EscapeUriString(refId));

            if (parameters.Age > 0)
                data.Append("&age=").Append(Uri.EscapeUriString(parameters.Age.ToString(parameters.culture)));
            data.Append("&altitude=").Append(Uri.EscapeUriString(parameters.Altitude.ToString()));
            if (parameters.EventContentType != null)
                data.Append("&content_type=").Append(Uri.EscapeUriString(parameters.EventContentType));
            if (parameters.EventContentId != null)
                data.Append("&content_id=").Append(Uri.EscapeUriString(parameters.EventContentId));
            data.Append("&level=").Append(parameters.EventLevel.ToString());
            data.Append("&quantity=").Append(parameters.EventQuantity.ToString());
            if (parameters.EventSearchString != null)
                data.Append("&search_string=").Append(Uri.EscapeUriString(parameters.EventSearchString));
            data.Append("&rating=").Append(Uri.EscapeUriString(parameters.EventRating.ToString()));
            if (parameters.EventDate1 != null)
                data.Append("&date1=").Append(Uri.EscapeUriString(UnixTimestamp(parameters.EventDate1).ToString()));
            if (parameters.EventDate2 != null)
                data.Append("&date2=").Append(Uri.EscapeUriString(UnixTimestamp(parameters.EventDate2).ToString()));
            if (parameters.EventAttribute1 != null)
                data.Append("&attribute_sub1=").Append(Uri.EscapeUriString(parameters.EventAttribute1));
            if (parameters.EventAttribute2 != null)
                data.Append("&attribute_sub2=").Append(Uri.EscapeUriString(parameters.EventAttribute2));
            if (parameters.EventAttribute3 != null)
                data.Append("&attribute_sub3=").Append(Uri.EscapeUriString(parameters.EventAttribute3));
            if (parameters.EventAttribute4 != null)
                data.Append("&attribute_sub4=").Append(Uri.EscapeUriString(parameters.EventAttribute4));
            if (parameters.EventAttribute5 != null)
                data.Append("&attribute_sub5=").Append(Uri.EscapeUriString(parameters.EventAttribute5));
            if (parameters.DeviceBrand != null)
                data.Append("&device_brand=").Append(Uri.EscapeUriString(parameters.DeviceBrand));
            if (parameters.DeviceModel != null)
                data.Append("&device_model=").Append(Uri.EscapeUriString(parameters.DeviceModel));
            if (parameters.DeviceType != null)
                data.Append("&device_type=").Append(Uri.EscapeUriString(parameters.DeviceType));
            if (parameters.FacebookUserId != null)
                data.Append("&facebook_user_id=").Append(Uri.EscapeUriString(parameters.FacebookUserId));
            if (parameters.Gender != MATGender.NONE)
                data.Append("&gender=").Append(Uri.EscapeUriString(((int)parameters.Gender).ToString(parameters.culture)));
            if (parameters.GoogleUserId != null)
                data.Append("&google_user_id=").Append(Uri.EscapeUriString(parameters.GoogleUserId));
            if (parameters.IsPayingUser != false)
                data.Append("&is_paying_user=1");
            if (parameters.Latitude != 0)
                data.Append("&latitude=").Append(Uri.EscapeUriString(parameters.Latitude.ToString(parameters.culture)));
            if (parameters.Longitude != 0)
                data.Append("&longitude=").Append(Uri.EscapeUriString(parameters.Longitude.ToString(parameters.culture)));
            if (parameters.TwitterUserId != null)
                data.Append("&twitter_user_id=").Append(Uri.EscapeUriString(parameters.TwitterUserId));
            if (parameters.UserEmail != null)
                data.Append("&user_email=").Append(Uri.EscapeUriString(parameters.UserEmail));
            if (parameters.UserId != null)
                data.Append("&user_id=").Append(Uri.EscapeUriString(parameters.UserId));
            if (parameters.UserName != null)
                data.Append("&user_name=").Append(Uri.EscapeUriString(parameters.UserName));

            // Add event items to url as json string
            if (eventItems != null)
                data.Append("&site_event_items=").Append(Uri.EscapeUriString(JsonConvert.SerializeObject(eventItems)));

            if (parameters.matRequest != null)
                parameters.matRequest.ParamsToBeEncrypted(data.ToString());

            // Encrypt data string
            string dataStr = parameters.urlEncrypter.Encrypt(data.ToString());
            url.Append("&data=").Append(dataStr);

            url.Append("&response_format=json");

            if (parameters.matRequest != null)
                parameters.matRequest.ConstructedRequest(url.ToString());

            return url.ToString();
        }

        public static long UnixTimestamp()
        {
            return UnixTimestamp(DateTime.UtcNow);
        }

        public static long UnixTimestamp(DateTime? date)
        {
            var utcEpoch = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
            var span = date - utcEpoch;
            return (long)(span ?? TimeSpan.Zero).TotalSeconds;
        }
    }
}
